application-operator/mcagent/multicluster_agent.go

```
@@ -0,0 +1,161 @@
+		ClusterName: secret.ClusterName,
+		Context:     context.TODO(),
+	}
+
+	go s.StartSync()
+}
+
+// StartSync - start the thread for syncing multi-cluster objects
+func (s *Syncer) StartSync() {
+	s.Log.Info("Starting sync of multi-cluster objects")
+
+	// Periodically loop looking for multi-cluster objects
+	for {
+		err := s.syncMCSecretObjects()
+		if err != nil {
+			s.Log.Error(err, "Error syncing MultiClusterSecret objects")
+		}
+		err = s.syncMCConfigMapObjects()
+		if err != nil {
+			s.Log.Error(err, "Error syncing MultiClusterConfigMap objects")
+		}
+		err = s.syncMCLoggingScopeObjects()
+		if err != nil {
+			s.Log.Error(err, "Error syncing MultiClusterLoggingScope objects")
+		}
+		err = s.syncMCComponentObjects()
+		if err != nil {
+			s.Log.Error(err, "Error syncing MultiClusterComponent objects")
+		}
+		err = s.syncMCApplicationConfigurationObjects()
+		if err != nil {
+			s.Log.Error(err, "Error syncing MultiClusterApplicationConfiguration objects")
+		}
+		time.Sleep(5 * time.Minute)
+	}
+}
+
+// Validate the cluster secret
+func validateClusterSecret(secret *corev1.Secret) error {
+	// The secret must contain a cluster-name
+	_, ok := secret.Data["cluster-name"]
+	if !ok {
+		return fmt.Errorf("the secret named %s in namespace %s is missing the required field cluster-name", secret.Name, secret.Namespace)
+	}
+
+	// The secret must contain a kubeconfig
+	_, ok = secret.Data["kubeconfig"]
+	if !ok {
+		return fmt.Errorf("the secret named %s in namespace %s is missing the required field kubeconfig", secret.Name, secret.Namespace)
+	}
+
+	return nil
+}
+
+// Get the clientset for accessing the admin cluster
+func getAdminClient(secret *corev1.Secret) (client.Client, error) {
+	// Create a temp file that contains the kubeconfig
+	tmpFile, err := ioutil.TempFile("/tmp", "kubeconfig")
+	if err != nil {
+		return nil, err
+	}
+
+	err = ioutil.WriteFile(tmpFile.Name(), secret.Data["kubeconfig"], 0777)
+	defer os.Remove(tmpFile.Name())
```

maybe defer should be after the error check

---

I don't know but just in case it seems safer to have it before the error check.  Worst case is the remove fails silently.